File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class BSTIterator {
    IList<int> data = new List<int>();
    int pointer = -1;
    
    public BSTIterator(TreeNode root) {
        this.data = this.inorder(root);
    }
    
    public int Next() {
        return this.data[++this.pointer];
    }
    
    public bool HasNext() {
        return this.pointer < this.data.Count - 1;
    }
    
    private IList<int> inorder(TreeNode root) {
        IList<int> data = new List<int>();
        
        void dfs(TreeNode node) {
            if (node.left != null) dfs(node.left);
            data.Add(node.val);
            if (node.right != null) dfs(node.right);
        }
        
        dfs(root);
        return data;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */

File: medium/0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
class BSTIterator {
    constructor(root) {
        this.data = this.inorder(root);
        this.pointer = -1;
    }

    next() {
        this.pointer++;
        return this.data[this.pointer];
    }

    hasNext() {
        return this.pointer < this.data.length - 1;
    }

    inorder(root) {
        const data = [];

        const dfs = node => {
            if (!node) return;

            dfs(node.left);
            data.push(node.val);
            dfs(node.right);
        }

        dfs(root);

        return data;
    }
};



/** 
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */